### PR TITLE
Convert addon to V2

### LINKED ISSUE #217: Convert addon to v2 & allow older ember-source versions than 4.12
While updating `ember-basic-dropdown` to a V2 addon, i have seen that this addon is atm a V1 addon.
I would like that `ember-basic-dropdown` brings early / later all dependencies as addon V2 (better earlier 😅).

Is there planned to port this addon to V2? I think that its very easy for this addon, because the addon is very lite.
The requirement for V2 addon is `ember-auto-import` v2 or embroider (see [Addon V2 RFCS](https://rfcs.emberjs.com/id/0507-embroider-v2-package-format/))

The other idea is to allow also older `ember-source` (atm v4.12+ is required). This means that only latest 2 LTS versions are supported. I don't understand why this is necessary... Is there any reason?
All dependency packages are allowing at least v3.28+... so would be great that we can also support this addon for older ember apps.

If you need any help to convert to V2 please let me know.

### PR DESCRIPTION
**Pull request by mkszepp**
Fix #217

- Update addon to V2
- Allow ember-source `3.28`, `4.x` and `>5.0`

### CODE REVIEW

**Comment by mkszepp** [comment]
@jelhan i think this is ready... please approve workflow, so we can check of all tests do pass

**Comment by jelhan** [comment]
Thanks a lot for working on picking this up that quickly!

How did you converted it? Is it based on [Embroider addon blueprints](https://github.com/embroider-build/addon-blueprint)?

I noticed some misalignments with Ember CLI conventions. E.g. Prettier configuration is different. If this mismatch is caused by the blueprints, I would report upstream rather than discussing in this PR. That's why I'm asking.

**Comment by mkszepp** [comment]
@jelhan for v2 addon migration i have always used https://github.com/NullVoxPopuli/ember-addon-migrator this is using embroider blueprint internal

The migrator has some problems, like readme, changelog or renovate will not be readded correctly. This is a manual step for the moment.
Also while migration i have got errors by porting other addons (this one was going very simple).

The migrator make +- new apps and is readding all files from before.

I will check the prettier file...

**Comment by jelhan** [comment]
Thanks a lot for providing the details. If changes are due to different opinions of Ember CLI and Embroider blueprints that's fine for me and nothing which needs to be changed here. I will report those upstream and of there is a decision to align, it should propagate to this repo in next Ember CLI Upgrade.

Prettier is one example. Ember CLI configures Prettier to use double quotes as standard. And only overwrites for JavaScript and TypeScript files to use single quotes instead. ([Source](https://github.com/ember-cli/ember-cli/blob/f824a6aee6e804fe4584fb037e893e3bbf93370a/blueprints/app/files/.prettierrc.js#L8C2-L8C2)) Embroider addon blueprints configure Prettier to use single quotes for all files. ([Source](https://github.com/embroider-build/addon-blueprint/blob/b1168b4616adf748bd713c5af39e104325982e9b/files/.prettierrc.cjs#L5)) For this PR both is fine for me. But I think long-term it should be aligned between the blueprints.

I hope I have some time doing a full review this weekend. Otherwise I will get back to it early next week.

**Comment by mkszepp** [comment]
@jelhan did you have time for review?

File: test-app/types/ember-data/types/registries/model.d.ts
Comment (jelhan) [inline test-app/types/ember-data/types/registries/model.d.ts:1]: I don't think this is needed as we don't use ember-data.

File: test-app/package.json
Comment (jelhan) [inline test-app/package.json:81]: I don't think we need ember-data here.

File: test-app/package.json
Comment (jelhan) [inline test-app/package.json:44]: Types for ember-data shouldn't be needed as ember-data is not used.

File: test-app/package.json
Comment (jelhan) [inline test-app/package.json:112]: I don't think we need that dependency.

File: test-app/package.json
Comment (jelhan) [inline test-app/package.json:114]: The dependencies aren't ordered alphabetically. That will cause unnecessary churn when PNPM or RenovateBot reorders them. Could you please fix?

File: test-app/package.json
Comment (jelhan) [inline test-app/package.json:124]: Why is this a dependency and not a dev dependency?

**Review by jelhan — CHANGES_REQUESTED** [review]
Thanks for the reminder. I'm sorry for not getting back to this earlier. I didn't had much time for open source recently. And reviewing the v2 conversion was a rather complex task due to amount of changes.

File: .github/workflows/ci.yml
Comment (mkszepp) [reply .github/workflows/ci.yml:20]: just blueprint, because he brings as default atm node v16. We should hold v18, because v16 is not anymore LTS

File: .github/workflows/ci.yml
Comment (mkszepp) [reply .github/workflows/ci.yml:23]: v18 is correct, i will updated with next commit

File: ember-style-modifier/.prettierrc.cjs
Comment (mkszepp) [reply ember-style-modifier/.prettierrc.cjs]: this brings the blueprint... i will try to remove and check if it works when we have on root

File: test-app/.editorconfig
Comment (mkszepp) [reply test-app/.editorconfig:1]: root its ok, removed on test-app (happend while migration)

File: test-app/config/ember-cli-update.json
Comment (mkszepp) [reply test-app/config/ember-cli-update.json:16]: --ci-provider=travis is definitely wrong idk when it was added, because `--ci-provider=travis` is deprecated... i have replaced like there was before

File: test-app/types/ember-data/types/registries/model.d.ts
Comment (mkszepp) [reply test-app/types/ember-data/types/registries/model.d.ts:1]: aggr.. added by blueprint and i have forgot to remove... sry

File: test-app/package.json
Comment (mkszepp) [reply test-app/package.json:81]: aggr.. added by blueprint and i have forgot to remove... sry

File: test-app/package.json
Comment (mkszepp) [reply test-app/package.json:124]: it was added from blueprint / migration tool... i think correctly is as devDepedency

File: ember-style-modifier/src/types/global.d.ts
Comment (mkszepp) [reply ember-style-modifier/src/types/global.d.ts:1]: its necessary that is working `glint` https://github.com/typed-ember/glint

File: package.json
Comment (mkszepp) [reply package.json:9]: filter is the standard from blueprint (which are using also other addons)... --filter tells to use only a specific folder --recursive says every folder but we need to build only the addon folder

**Comment by mkszepp** [comment]
@jelhan please don't merge renotate bot changes atm... because its not possible to merge this changes. i have now updated manually the merged dependency updates

File: ember-style-modifier/src/template-registry.ts
Comment (mkszepp) [reply ember-style-modifier/src/template-registry.ts:1]: @jelhan i'm not 100% sure, but i have seen that this files will be exported always empty in other addons...

The glint docs says, that the registry should be added directly in helper / component or?
https://typed-ember.gitbook.io/glint/environments/ember/template-registry#helpers-and-modifiers

Comment (jelhan) [reply ember-style-modifier/src/template-registry.ts:1]: This is the relevant docs for authoring addons: https://typed-ember.gitbook.io/glint/environments/ember/authoring-addons

We should provide a Template Registry, which consumers can import to register the helper. This is explained here: https://typed-ember.gitbook.io/glint/environments/ember/using-addons#using-glint-enabled-addons

Unless there is some magic in the v2 addon blueprint to generate this file automatically, it should be still needed.

I'm only confused that the import changed from type-only to regular import. That I don't get.

Comment (jelhan) [reply ember-style-modifier/src/template-registry.ts:1]: I just noticed that the template registry is not consumed in `test-app/types/global.d.ts`. I'm confused. How does the `{{style}}` modifier gets registered in test-app so that glint knows the types for it?

File: ember-style-modifier/src/types/global.d.ts
Comment (jelhan) [reply ember-style-modifier/src/types/global.d.ts:1]: I still don't get it. It should be necessary if we have any loose-mode template, which uses `{{style}}` modifier. But we don't have that in the addon. We only have it in the test app.

I would have expected this in test-app but not in the addon. But it seems to be exactly the other way around. I'm confused

File: ember-style-modifier/.prettierrc.cjs
Comment (jelhan) [reply ember-style-modifier/.prettierrc.cjs:1]: If it's coming from blueprint I'm fine with keeping it.

File: ember-style-modifier/src/template-registry.ts
Comment (mkszepp) [reply ember-style-modifier/src/template-registry.ts]: i have now changes the file like there was before... because this is correct.. in additional i have `addon.publicEntrypoints(['index.js', '**/*.js']),` with `addon.publicEntrypoints(['index.js', 'modifiers/**/*.js'])`, because `'**/*.js'` has created the empty `template-registry.js`. The used `template-registry.js` for glint is `declarations/template-registry.js` and this was exported in both ways correctly... but we should this what the docs says.

In test-app there was missing a part of `glint`... there was only ts check active in package.json... now everything is correctly in `test-app`

So i think, we have resolved this one

File: ember-style-modifier/src/types/global.d.ts
Comment (mkszepp) [reply ember-style-modifier/src/types/global.d.ts]: thats true... this is only necessary in test-app... i have removed now

File: .github/workflows/ci.yml
Comment (jelhan) [inline .github/workflows/ci.yml:40]: I think we need this because otherwise it is not testing with floating dependencies.

File: .github/workflows/ci.yml
Comment (jelhan) [inline .github/workflows/ci.yml:44]: Can we please reset to double quotes? This is how Prettier is configured for YAML files. Even though we don't check them in CI.

```suggestion
    needs: "test"
```

**Review by jelhan — CHANGES_REQUESTED** [review]
Thanks a lot for keeping working on this!

Seems all good beside one bug in GitHub Actions and one code formatting issue for YAML files.

CI pipeline is failing because one recently added file wasn't formatted with Prettier.

I haven't tested locally the created package yet. Will do that as a last step before merging.

**Comment by mkszepp** [comment]
@jelhan hope the test do pass now, otherwise let me know

**Comment by jelhan** [comment]
Released as [v4.2.0](https://github.com/jelhan/ember-style-modifier/releases/tag/v4.2.0).

**Pull request merged**
